### PR TITLE
NMS-14173: Add SCV Rest API

### PR DESCRIPTION
**Pull request by cgorantla**
### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-14173

### CODE REVIEW

File: features/scv/rest/src/main/java/org/opennms/features/scv/rest/DefaultScvRestService.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.features.scv.rest;
+
+import joptsimple.internal.Strings;
```
Comment (synqotik) [inline features/scv/rest/src/main/java/org/opennms/features/scv/rest/DefaultScvRestService.java:31]: Should be `com.google.common.base.Strings` ?

File: features/scv/rest/src/main/java/org/opennms/features/scv/rest/DefaultScvRestService.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+                    credentialsDTO.getPassword(), credentialsDTO.getAttributes());
+            try {
+                scv.setCredentials(credentialsDTO.getAlias(), credentials);
+                return Response.accepted().build();
+            } catch (Exception e) {
+                LOG.error("Exception while adding credentials with alias {} ", credentialsDTO.getAlias(), e);
+                return Response.status(Response.Status.INTERNAL_SERVER_ERROR).entity(e.getMessage()).build();
+            }
+        } else {
+            return Response.status(Response.Status.BAD_REQUEST).entity("Missing alias for credentails").build();
```
Comment (synqotik) [inline features/scv/rest/src/main/java/org/opennms/features/scv/rest/DefaultScvRestService.java:85]: `credentails` -> `credentials`

File: features/scv/rest/src/main/java/org/opennms/features/scv/rest/DefaultScvRestService.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    public Response editCredentials(String alias, CredentialsDTO credentialsDTO) {
+
+        if (credentialsDTO != null && !Strings.isNullOrEmpty(alias) && alias.equals(credentialsDTO.getAlias())) {
+            Credentials credentials = new Credentials(credentialsDTO.getUsername(),
+                    credentialsDTO.getPassword(), credentialsDTO.getAttributes());
+            try {
+                scv.setCredentials(alias, credentials);
+                return Response.accepted().build();
+            } catch (Exception e) {
+                LOG.error("Exception while adding credentials with alias {} ", credentialsDTO.getAlias(), e);
```
Comment (synqotik) [inline features/scv/rest/src/main/java/org/opennms/features/scv/rest/DefaultScvRestService.java:99]: Remove trailing space in message: `alias {} "` -> `alias {}"`

File: features/scv/rest/src/main/java/org/opennms/features/scv/rest/DefaultScvRestService.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+                return Response.status(Response.Status.INTERNAL_SERVER_ERROR).entity(e.getMessage()).build();
+            }
+        } else {
+            return Response.status(Response.Status.BAD_REQUEST).build();
+        }
+    }
+
+    @Override
+    public Response getAliases() {
+        Set<String> res = scv.getAliases();
```
Comment (synqotik) [inline features/scv/rest/src/main/java/org/opennms/features/scv/rest/DefaultScvRestService.java:109]: Should these be sorted alphabetically?

**Review by synqotik — APPROVED** [review]
Thanks for changes, LGTM!

File: features/scv/rest/src/main/java/org/opennms/features/scv/rest/CredentialsDTO.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    private String alias;
+
+    private String username;
+    private String password;
+    private Map<String, String> attributes;
+
+    public CredentialsDTO() {
+    }
+
+    public CredentialsDTO(String alias, String username, String password) {
```
Comment (christianpape) [inline features/scv/rest/src/main/java/org/opennms/features/scv/rest/CredentialsDTO.java:45]: Maybe include attributes in the constructor?

**Review by christianpape — DISMISSED** [review]
LGTM.

File: features/scv/rest/src/main/java/org/opennms/features/scv/rest/CredentialsDTO.java (hunk 5 above)
Comment (cgorantla) [reply features/scv/rest/src/main/java/org/opennms/features/scv/rest/CredentialsDTO.java:46]: Done!

File: features/scv/rest/src/main/java/org/opennms/features/scv/rest/DefaultScvRestService.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+    @Override
+    public Response getCredentials(String alias) {
+        try {
+            Credentials credentials = scv.getCredentials(alias);
+            if (credentials == null) {
+                return Response.noContent().build();
+            }
+            if (credentials.getUsername() != null && credentials.getPassword() != null) {
+                CredentialsDTO dto = new CredentialsDTO(alias, credentials.getUsername(), credentials.getPassword());
```
[the remark this answers is not in the document]
Comment (cgorantla) [reply features/scv/rest/src/main/java/org/opennms/features/scv/rest/DefaultScvRestService.java]: It was masked in the next line but moved to constructor now

**Review by synqotik — APPROVED** [review]
Looks good

**Pull request merged**